core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala

```diff
@@ -0,0 +1,746 @@
+import org.broadinstitute.dsde.rawls.monitor.migration.MigrationUtils._
+import org.broadinstitute.dsde.rawls.monitor.migration.WorkspaceMigrationHistory._
+import org.broadinstitute.dsde.rawls.workspace.WorkspaceService
+import org.broadinstitute.dsde.workbench.google2.GoogleStorageTransferService.JobTransferSchedule
+import org.broadinstitute.dsde.workbench.google2.{GoogleStorageService, GoogleStorageTransferService}
+import org.broadinstitute.dsde.workbench.model.google.{GcsBucketName, GoogleProject}
+
+import java.sql.Timestamp
+import java.time.{Instant, LocalDateTime, ZoneOffset}
+import java.util.UUID
+import scala.collection.JavaConversions._
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration.FiniteDuration
+import scala.language.higherKinds
+
+
+object WorkspaceMigrationActor {
+
+  import slick.jdbc.MySQLProfile.api._
+
+  val workspaceMigrations = WorkspaceMigrationHistory.workspaceMigrations
+  val storageTransferJobs = PpwStorageTransferJobs.storageTransferJobs
+
+
+  final def truncate: WriteAction[Unit] =
+    storageTransferJobs.delete >> workspaceMigrations.delete >> DBIO.successful()
+
+
+  final def isInQueueToMigrate(workspace: Workspace): ReadWriteAction[Boolean] =
+    workspaceMigrations
+      .filter { m => m.workspaceId === workspace.workspaceIdAsUUID && !m.started.isDefined }
+      .length
+      .result
+      .map(_ > 0)
+
+
+  final def isMigrating(workspace: Workspace): ReadWriteAction[Boolean] =
+    workspaceMigrations
+      .filter { m =>
+        m.workspaceId === workspace.workspaceIdAsUUID &&
+          m.started.isDefined &&
+          m.finished.isEmpty
+      }
+      .length
+      .result
+      .map(_ > 0)
+
+
+  final def schedule(workspace: Workspace): ReadWriteAction[Unit] =
+    workspaceMigrations
+      .map(_.workspaceId)
+      .insert(workspace.workspaceIdAsUUID)
+      .ignore
+
+
+  final case class MigrationDeps(dataSource: SlickDataSource,
+                                 googleProjectToBill: GoogleProject,
+                                 workspaceService: WorkspaceService,
+                                 storageService: GoogleStorageService[IO],
+                                 storageTransferService: GoogleStorageTransferService[IO])
+
+
+  type MigrateAction[A] = ReaderT[OptionT[IO, *], MigrationDeps, A]
+
+  object MigrateAction {
+
+    // lookup a value in the environment using `selector`
+    final def asks[A](selector: MigrationDeps => A): MigrateAction[A] =
+      ReaderT.ask[OptionT[IO, *], MigrationDeps].map(selector)
+
+    // create a MigrateAction that ignores its input and returns the OptionT
+    final def liftF[A](optionT: OptionT[IO, A]): MigrateAction[A] =
+      ReaderT.liftF(optionT)
+
+    // lift an IO action into the context of a MigrateAction
+    final def liftIO[A](ioa: IO[A]): MigrateAction[A] =
+      ReaderT.liftF(OptionT.liftF(ioa))
```

Comment:
I don't understand the question - do you want me to call this function `liftF`? If that's the case then no.
`liftF` (or `lift`) comes from `MonadTrans` which can be written in scala as:
```scala
trait MonadTrans[T[_, _], F[_]] {
   def liftF[A](fa: F[A]): T[F, A]
}
```
`liftF` only lifts one level down. `liftIO` is a special kind of a monad transformer and has it's own type class because lifting IO is a common pattern:
```scala
trait LiftIO[F[_]] {
   def liftIO[A](ioa: IO[A]): F[A]
}
```
`liftIO` can lift an `IO` at any level (unlike `liftF` where it's one level). We're trying to "lift" an IO into `OptionT` and then into `ReaderT`, thus `liftIO` is appropriate. Does that help?

Reply:
Got it! Thank you!